scipy/sparse/csgraph/_laplacian.py

```diff
@@ -83,78 +121,222 @@ def laplacian(csgraph, normed=False, return_diag=False, use_out_degree=False,
+    if dtype is None:
+        dtype = graph.dtype
+
+    if form != "array":
+        graph_sum = graph.sum(axis=axis).getA1()
+        graph_diagonal = graph.diagonal()
+        diag = graph_sum - graph_diagonal
+        if symmetrized:
+            graph_sum += graph.sum(axis=1 - axis).getA1()
+            diag = graph_sum - graph_diagonal - graph_diagonal
+
+        if normed:
+            isolated_node_mask = diag == 0
+            w = np.where(isolated_node_mask, 1, np.sqrt(diag))
+            if symmetrized:
+                md = _md_normed_sym(graph, graph_sum, 1.0 / w)
+            else:
+                md = _md_normed(graph, graph_sum, 1.0 / w)
+            if form == "function":
+                return md, w.astype(dtype, copy=False)
+            elif form == "lo":
+                m = _linearoperator(md, shape=graph.shape, dtype=dtype)
+                return m, w.astype(dtype, copy=False)
+        else:
+            if symmetrized:
+                md = _md_sym(graph, graph_sum)
+            else:
+                md = _md(graph, graph_sum)
+            if form == "function":
+                return md, diag.astype(dtype, copy=False)
+            elif form == "lo":
+                m = _linearoperator(md, shape=graph.shape, dtype=dtype)
+                return m, diag.astype(dtype, copy=False)
+
     else:
-        if m.format == 'dia':
-            m = m.copy()
+        needs_copy = False
+        if graph.format in ("lil", "dok"):
+            m = graph.tocoo()
         else:
+            m = graph
+            if copy:
+                needs_copy = True
+
+        if symmetrized:
+            m += m.T.conj()
+
+        w = m.sum(axis=axis).getA1() - m.diagonal()
+        if normed:
             m = m.tocoo(copy=needs_copy)
-        m.data *= -1
-        m.setdiag(w)
-    return m, w
+            isolated_node_mask = w == 0
+            w = np.where(isolated_node_mask, 1, np.sqrt(w))
+            m.data /= w[m.row]
+            m.data /= w[m.col]
+            m.data *= -1
+            m.setdiag(1 - isolated_node_mask)
+        else:
+            if m.format == "dia":
+                m = m.copy()
+            else:
+                m = m.tocoo(copy=needs_copy)
+            m.data *= -1
+            m.setdiag(w)
+
+        return m.astype(dtype, copy=False), w.astype(dtype)
 
 
-def _laplacian_dense(graph, normed=False, axis=0,
-                     copy=True):
+def _laplacian_dense(graph, normed, axis, copy, form, dtype, symmetrized):
+
+    if dtype is None:
+        dtype = graph.dtype
+
     if copy:
         m = np.array(graph)
     else:
         m = np.asarray(graph)
 
-    np.fill_diagonal(m, 0)
-    w = m.sum(axis=axis)
-    if normed:
-        isolated_node_mask = (w == 0)
-        w = np.where(isolated_node_mask, 1, np.sqrt(w))
-        m /= w
-        m /= w[:, np.newaxis]
-        m *= -1
-        _setdiag_dense(m, 1 - isolated_node_mask)
+    if dtype is None:
+        dtype = m.dtype
+
+    if form != "array":
```

Comment:
Same comment here about splitting this into two separate functions

Reply:
done